common/flatpak-run.c

```diff
@@ -434,13 +436,23 @@ flatpak_run_add_wayland_args (FlatpakBwrap *bwrap)
 
   sandbox_wayland_socket = g_strdup_printf ("/run/flatpak/%s", wayland_display);
 
+  wayland_socket_lock = g_strdup_printf ("%s.lock", wayland_socket);
+  sandbox_wayland_socket_lock = g_strdup_printf ("%s.lock", sandbox_wayland_socket);
+
   if (stat (wayland_socket, &statbuf) == 0 &&
       (statbuf.st_mode & S_IFMT) == S_IFSOCK)
     {
       res = TRUE;
       flatpak_bwrap_add_args (bwrap,
                               "--ro-bind", wayland_socket, sandbox_wayland_socket,
                               NULL);
+
+      if (g_file_test (wayland_socket_lock, G_FILE_TEST_EXISTS))
+        flatpak_bwrap_add_args (bwrap,
+                                "--ro-bind", wayland_socket_lock,
+                                sandbox_wayland_socket_lock,
+                                NULL);
+
       flatpak_bwrap_add_runtime_dir_member (bwrap, wayland_display);
```

Comment:
How do Wayland implementations find the `.lock` file? Do they look for it relative to the `realpath()` of the socket, or relative to the logical path of the socket, which is usually in `$XDG_RUNTIME_DIR`?

If they (sometimes or always) look for it relative to the logical path of the socket, you will need to call `flatpak_bwrap_add_runtime_dir_member()` again for `wayland-0.lock` (or whatever the filename is), inside the `g_file_test()` check, to create a symlink like `$XDG_RUNTIME_DIR/wayland-0.lock -> /run/flatpak/wayland-0.lock`.

The reason we do it like this is that `$XDG_RUNTIME_DIR` is at least partially application-controlled since #4126 (and in a sandboxing framework we should treat the application as at least partially hostile), whereas `/run/flatpak` is fully Flatpak-controlled. See #4126 for more details.

Reply:
Looking at `src/wayland-server.c` in wayland, it seems like it's the logical path that's used, so yes, the extra `flatpak_bwrap_add_runtime_dir_member()` call probably *is* necessary.